BXMSDOC-5387: Create process services doc for Kogito.

sterobin opened the pull request:
**Cristiano, Eder,** can you review and verify this new "Developing process services with Kogito" content for Dev Preview? It contains core BA 7.x BPMN content adapted and overhauled for Kogito, plus all the BPMN-related content from the Kogito runtimes wiki.

Cristiano, see the few questions tagged `@Cristiano` in the doc preview.

See doc previews etc. below.

Links:
* [JIRA](https://issues.redhat.com/browse/BXMSDOC-5375)
* [Enterprise doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5387/)
* [Community doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5387_COMM/#chap_kogito-developing-process-services) - Same, but conditioned accordingly

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+
+* Start events to initiate the process
+* Tasks or other steps that are completed as part of the process
+* Connectors to link the process nodes and create a flow chart
```
perhaps, create a/the sequence flow?

sterobin replied:
Yes, better. Done.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+The `Person.bpmn2` process invokes a Decision Model and Notation (DMN) model in a business rule task to determine customer age, followed by a user task for special handling requirements for children, if applicable.
+
+.Example `Person.bpmn2` process invoked as a subprocess
+image::kogito/bpmn/bpmn-model-example-evaluation.png[Image of `OrderItems.bpmn` example process]
```
@sterobin this image is from another process, please review.

sterobin replied:
Ha! I noticed it yesterday when reusing the same in another and hoped to correct it before you saw :) Sorry. Fixed.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+.Example `Person.bpmn2` process invoked as a subprocess
+image::kogito/bpmn/bpmn-model-example-evaluation.png[Image of `OrderItems.bpmn` example process]
+
+Based on these processes and on application configurations, this example service exposes REST operations to create new orders, to list and delete active orders, and to determine the age of any given person.
```
```suggestion
Based on these processes and on application configurations, {PRODUCT} generates a set of REST operations to create new orders, to list and delete active orders, and to determine the age of any given person.
```

sterobin replied:
Done.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+.Example curl command to retrieve an order by ID
+[source]
+----
+curl -X GET http://localhost:8080/orders/12345
```
@sterobin should it be more explicit that this is the process id returned by the post or get command?  Just want to make sure people don't get confused with _orderNumber_.

sterobin replied:
Oops, yes, should. I clarified this with the same example in the other "Creating and running" doc. Have updated and clarified these endpoint examples here. Done.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+.Example curl command to cancel an order by ID
+[source]
+----
+curl -X DELETE http://localhost:8080/orders/12345
```
Same here, regarding id.

sterobin replied:
Done.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-model-example.adoc:
```diff
@@ -0,0 +1,188 @@
+curl -X DELETE http://localhost:8080/orders/12345
+----
+
+The following is the BPMN source file for `Orders.bpmn2` process model, as an example:
```
To be honest, I don't see much value in showing the raw xml.

sterobin replied:
Well, this section serves as a guided tour of a BPMN model example, so just showing what drives it. We do the same with the DMN doc content (show a DMN XML file as an illustration), so am just trying to keep things consistent.

cristianonicolai replied:
yep. that's fine, thanks.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-support.adoc:
```diff
@@ -0,0 +1,110 @@
+[id='ref_bpmn-support_{context}']
+= BPMN2 support in {PRODUCT}
+
+{PRODUCT} currently supports a subset of the https://www.omg.org/spec/BPMN/2.0/About-BPMN[Business Process Model and Notation 2.0 specification]. Although the {PRODUCT} BPMN designer displays many BPMN components in the canvas palette, the {PROCESS_ENGINE} (process runtime component) in {PRODUCT} currently executes only the subset of components. If you use any BPMN components from the {PRODUCT} B ...
```
{PROCESS_ENGINE} is showing as jBPM. I think we should avoid that. should be just about Kogito in here, so replacing with {PRODUCT}?

sterobin replied:
Ah, yes, the distinction of "decision/Drools engine" and "process/jBPM engine" in enterprise vs. community came from BA 7.x. I kept only until we had a decision, which we now have as of yesterday. Mark P., Alessandro, and Edoardo V. determined that for Kogito, we're just going with "decision engine" and "process engine". I will continue to use the variables for consistency and convenience if/when we want to change it again (best practice to maintain standard component names as variables). So I'll update the variable definition now for community. Thanks.

cristianonicolai replied:
perfect, thanks

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-tasks.adoc:
```diff
@@ -0,0 +1,261 @@
+
+In addition, the BPMN2 specification provides the ability to create custom tasks. The following predefined custom tasks are included with {PRODUCT}:
+
+* Rest service tasks: Used to invoke a remote RESTful service
```
should be removed as it is none of these are available in VsCode

sterobin replied:
Done

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_persistence-enabling.adoc:
```diff
@@ -0,0 +1,64 @@
+You can enable persistence in your {PRODUCT} services to persist data, such as active process nodes and process instance variables, so that the data is preserved across application restarts. {PRODUCT} persistence is based on https://infinispan.org/[Infinispan], so you must have a an Infinispan server installed and running in order to enable persistence.
+
+.Prerequisites
+* https://infinispan.org/[Infinispan] server is installed and running. For information about Infinispan installation and configuration, see the https://infinispan.org/documentation/[Infinispan documentation].
```
Perhaps good to mentioned Inifinispan 10+

sterobin replied:
Okay, done.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_persistence-enabling.adoc:
```diff
@@ -0,0 +1,64 @@
+<dependency>
+  <groupId>org.kie.kogito</groupId>
+  <artifactId>infinispan-persistence-addon</artifactId>
+  <version></version>
```
```suggestion
  <version>${{PRODUCT_INIT}.version}</version>
```

sterobin replied:
Oops. thanks. Fixed.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_persistence-enabling.adoc:
```diff
@@ -0,0 +1,64 @@
+
+For example {PRODUCT} services with Infinispan persistence, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-infinispan-persistence-quarkus[`kogito-infinispan-persistence-quarkus`]: Example on Quarkus
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-infinispan-persistence-quarkus[`kogito-infinispan-persistence-quarkus`]: Example on Quarkus
```

sterobin replied:
Oh, strange. Not even sure how I ended up on master since stable is default. Fixed everywhere.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_persistence-enabling.adoc:
```diff
@@ -0,0 +1,64 @@
+For example {PRODUCT} services with Infinispan persistence, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-infinispan-persistence-quarkus[`kogito-infinispan-persistence-quarkus`]: Example on Quarkus
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-infinispan-persistence-springboot[`kogito-infinispan-persistence-springboot`]: Example on Spring Boot
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-infinispan-persistence-springboot[`kogito-infinispan-persistence-springboot`]: Example on Spring Boot
```

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_messaging-enabling.adoc:
```diff
@@ -0,0 +1,153 @@
+.Example message configuration for end event
+image::kogito/bpmn/bpmn-messaging-end-event.png[Image of message end event data]
+
+{PRODUCT} messaging is based on https://kafka.apache.org/[Apache Kafka], so you must have Kafka installed in order to enable messaging. Your marshalling configuration depends on the messaging solution you use.
```
We actually support MicroProfile Reactive Messaging, so that means that other messaging providers could be used. Perhaps rephrase that as, in this example we use Kafka, etc...

sterobin replied:
Unclear on this because the following was the desired phrasing by Duncan and Ricardo in the Kogito on OCP doc:

```
{PRODUCT} supports the MicroProfile Reactive Messaging specification for messaging in your services. {PRODUCT} messaging is based on https://kafka.apache.org/[Apache Kafka], so you must have Kafka installed in order to enable messaging.
```

I'll ping the three of you in gchat about this. Thanks.

sterobin replied:
Okay, confirmed the following wording with Ricardo in chat, with you pinged:

For this doc:

```
{PRODUCT} supports the MicroProfile Reactive Messaging specification for messaging in your services. ... For this procedure, the messaging is based on https://kafka.apache.org/[Apache Kafka], so you must have Kafka installed in order to enable messaging. Your marshalling configuration depends on the messaging solution you use.
```

For the OCP content, where Operator messaging is indeed based on Kafka:

```
{PRODUCT} supports the MicroProfile Reactive Messaging specification for messaging in your services. {PRODUCT} Operator messaging is based on https://kafka.apache.org/[Apache Kafka] and enables you to configure messages as either input or output of business process execution.
```

cristianonicolai replied:
looks good, thanks.

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_messaging-enabling.adoc:
```diff
@@ -0,0 +1,153 @@
+
+For example {PRODUCT} services with Kafka messaging, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-kafka-quickstart-quarkus[`kogito-kafka-quickstart-quarkus`]: Example on Quarkus
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-kafka-quickstart-quarkus[`kogito-kafka-quickstart-quarkus`]: Example on Quarkus
```

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_messaging-enabling.adoc:
```diff
@@ -0,0 +1,153 @@
+For example {PRODUCT} services with Kafka messaging, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-kafka-quickstart-quarkus[`kogito-kafka-quickstart-quarkus`]: Example on Quarkus
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-kafka-quickstart-quarkus[`kogito-kafka-quickstart-springboot`]: Example on Spring Boot
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-kafka-quickstart-quarkus[`kogito-kafka-quickstart-springboot`]: Example on Spring Boot
```

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_task-lifecycle-custom.adoc:
```diff
@@ -0,0 +1,81 @@
+
+For example {PRODUCT} services with custom task life cycle configurations, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-usertasks-custom-lifecycle-quarkus[`kogito-usertasks-custom-lifecycle-quarkus`]: Example on Quarkus
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-usertasks-custom-lifecycle-quarkus[`kogito-usertasks-custom-lifecycle-quarkus`]: Example on Quarkus
```

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_task-lifecycle-custom.adoc:
```diff
@@ -0,0 +1,81 @@
+For example {PRODUCT} services with custom task life cycle configurations, see the following example applications in GitHub:
+
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-usertasks-custom-lifecycle-quarkus[`kogito-usertasks-custom-lifecycle-quarkus`]: Example on Quarkus
+* https://github.com/kiegroup/kogito-examples/tree/master/kogito-usertasks-custom-lifecycle-springboot[`kogito-usertasks-custom-lifecycle-springboot`]: Example on Spring Boot
```
```suggestion
* https://github.com/kiegroup/kogito-examples/tree/stable/kogito-usertasks-custom-lifecycle-springboot[`kogito-usertasks-custom-lifecycle-springboot`]: Example on Spring Boot
```

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_task-life-cycle.adoc:
```diff
@@ -0,0 +1,119 @@
+
+NOTE: Authorization is only enforced when the method that calls the work item life cycle methods comes with a security context. This security behavior is dependent on the API that you use.
+
+== API interaction with task life cycle phases
```
@sterobin we should at least mention the REST endpoints here as another way to interact with the tasks and show the endpoint that are created. I belive you covered the content in some examples already.

sterobin replied:
I see. I don't think I get into any example endpoints elsewhere specifically for tasks, but this is my attempt using one of the [orderItems.bpmn2](https://github.com/kiegroup/kogito-examples/blob/stable/jbpm-quarkus-example/src/main/resources/org/kie/kogito/examples/orderItems.bpmn) supported endpoints, which I've added to the end of this section. 

**However, I can't get this to work, so clearly something is wrong. Can you see what I'm missing?**

```
The following REST endpoint example sets a task phase, group, and user in an `orderItems.bpmn2` process for verifying customer orders:

.Example POST endpoint to set a task phase, group, and user
[source]
----
http://localhost:8080/orderItems/__PROCESS_ID__/Verify_order/__WORK_ITEM_ID__?phase=__PHASE_NAME__&group=__GROUP_NAMES__&user=__USER_NAME__
----

.Example curl command to claim a task for a specified group and user
[source]
----
curl -X POST "http://localhost:8080/orderItems/demo.orderItems/Verify_order/UserTask_1?phase=claim&group=managers&user=john" -H "accept: application/json" -H "content-type: application/json"
----

.Example curl command to complete a task for a specified group and user
[source]
----
curl -X POST "http://localhost:8080/orderItems/demo.orderItems/Verify_order/UserTask_1?phase=complete&group=managers&user=john" -H "accept: application/json" -H "content-type: application/json"
----
```

cristianonicolai replied:
@sterobin should be a lot simpler than that, see https://github.com/kiegroup/kogito-examples/tree/stable/jbpm-quarkus-example#get-orderitemsidtasks.
Once you know the process id, you only need to do a get to retrieve the open tasks and a post to complete the task using both process id and task id as parameter.

cristianonicolai commented:
@sterobin thanks, this is really good. Two general notes, to review links point to examples to always use the _stable_ branch. Next one is regarding the use of Drools and jBPM engine, I believe we should avoid that and only mention as Kogito engine.

sterobin commented:
@cristianonicolai , thank you! I've implemented and addressed your general and specific feedback, but have left two remaining comments/questions for you above. Let me know what you think. Thanks.
